### PR TITLE
fix(datastore): Regression bug caused by changing the public enum value

### PR DESCRIPTION
**Pull request by royjit**
*Issue:* https://github.com/aws-amplify/amplify-ios/issues/882
*Description of changes:*  Regression bug caused by changing the public enum value

- Added static functions inside enum to match the previous enum cases to make it backward compatible.
- Unit tests for the changes.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: Amplify/Categories/DataStore/Model/Internal/Schema/ModelSchema+Definition.swift [hunk 1]
```diff
@@ -26,6 +26,14 @@ public enum ModelFieldType {
     case model(name: ModelName)
     case collection(of: ModelName)
 
+    public static func model(type: Model.Type) -> ModelFieldType {
+        return .model(name: type.modelName)
```
Comment (palpatim) [inline Amplify/Categories/DataStore/Model/Internal/Schema/ModelSchema+Definition.swift:30]: nit: remove redundant `return`
```suggestion
        .model(name: type.modelName)
```

File: Amplify/Categories/DataStore/Model/Internal/Schema/ModelSchema+Definition.swift [hunk 2]
```diff
@@ -26,6 +26,14 @@ public enum ModelFieldType {
...
+    }
+
+    public static func collection(of type: Model.Type) -> ModelFieldType {
+        return .collection(of: type.modelName)
```
Comment (palpatim) [inline Amplify/Categories/DataStore/Model/Internal/Schema/ModelSchema+Definition.swift:34]: nit: remove redundant `return`
```suggestion
        .collection(of: type.modelName)
```

File: AmplifyPlugins/Core/AWSPluginsCore/Model/GraphQLRequest/GraphQLRequest+AnyModelWithSync.swift [hunk 3]
```diff
@@ -64,6 +64,39 @@ extension GraphQLRequest: ModelSyncGraphQLRequestFactory {
                                                    decodePath: document.name)
     }
 
+    public static func createMutation(of model: Model, version: Int?) -> GraphQLRequest<MutationSyncResult> {
+        return createMutation(of: model, modelSchema: model.schema, version: version)
```
Comment (palpatim) [inline AmplifyPlugins/Core/AWSPluginsCore/Model/GraphQLRequest/GraphQLRequest+AnyModelWithSync.swift:68]: nit: remove redundant `return`
```suggestion
        createMutation(of: model, modelSchema: model.schema, version: version)
```

File: AmplifyPlugins/Core/AWSPluginsCore/Model/GraphQLRequest/GraphQLRequest+AnyModelWithSync.swift [hunk 4]
```diff
@@ -64,6 +64,39 @@ extension GraphQLRequest: ModelSyncGraphQLRequestFactory {
...
+    }
+
+    public static func updateMutation(of model: Model,
+                                      where filter: GraphQLFilter?,
+                                      version: Int?) -> GraphQLRequest<MutationSyncResult> {
+        return updateMutation(of: model, modelSchema: model.schema, where: filter, version: version)
```
Comment (palpatim) [inline AmplifyPlugins/Core/AWSPluginsCore/Model/GraphQLRequest/GraphQLRequest+AnyModelWithSync.swift:74]: nit: remove redundant `return`
```suggestion
        updateMutation(of: model, modelSchema: model.schema, where: filter, version: version)
```

File: AmplifyPlugins/Core/AWSPluginsCore/Model/GraphQLRequest/GraphQLRequest+AnyModelWithSync.swift [hunk 5]
```diff
@@ -64,6 +64,39 @@ extension GraphQLRequest: ModelSyncGraphQLRequestFactory {
...
+    }
+
+    public static func subscription(to modelType: Model.Type,
+                                    subscriptionType: GraphQLSubscriptionType) -> GraphQLRequest<MutationSyncResult> {
+        return subscription(to: modelType.schema, subscriptionType: subscriptionType)
```
Comment (palpatim) [inline AmplifyPlugins/Core/AWSPluginsCore/Model/GraphQLRequest/GraphQLRequest+AnyModelWithSync.swift:79]: nit: remove redundant `return`
```suggestion
        subscription(to: modelType.schema, subscriptionType: subscriptionType)
```

File: AmplifyPlugins/Core/AWSPluginsCore/Model/GraphQLRequest/GraphQLRequest+AnyModelWithSync.swift [hunk 6]
```diff
@@ -64,6 +64,39 @@ extension GraphQLRequest: ModelSyncGraphQLRequestFactory {
...
+    }
+
+    public static func subscription(to modelType: Model.Type,
+                                    subscriptionType: GraphQLSubscriptionType,
+                                    claims: IdentityClaimsDictionary) -> GraphQLRequest<MutationSyncResult> {
+        return subscription(to: modelType.schema, subscriptionType: subscriptionType, claims: claims)
```
Comment (palpatim) [inline AmplifyPlugins/Core/AWSPluginsCore/Model/GraphQLRequest/GraphQLRequest+AnyModelWithSync.swift:85]: nit: remove redundant `return`
```suggestion
        subscription(to: modelType.schema, subscriptionType: subscriptionType, claims: claims)
```

File: AmplifyPlugins/Core/AWSPluginsCore/Model/GraphQLRequest/GraphQLRequest+AnyModelWithSync.swift [hunk 7]
```diff
@@ -64,6 +64,39 @@ extension GraphQLRequest: ModelSyncGraphQLRequestFactory {
...
+    }
+
+    public static func syncQuery(modelType: Model.Type,
+                                 where predicate: QueryPredicate?,
+                                 limit: Int?,
+                                 nextToken: String?,
+                                 lastSync: Int?) -> GraphQLRequest<SyncQueryResult> {
+        return syncQuery(modelSchema: modelType.schema,
```
Comment (palpatim) [inline AmplifyPlugins/Core/AWSPluginsCore/Model/GraphQLRequest/GraphQLRequest+AnyModelWithSync.swift:93]: nit: remove redundant `return`
```suggestion
        syncQuery(modelSchema: modelType.schema,
```

File: AmplifyPlugins/Core/AWSPluginsCore/Model/GraphQLRequest/GraphQLRequest+Model.swift [hunk 8]
```diff
@@ -149,6 +149,12 @@ extension GraphQLRequest: ModelGraphQLRequestFactory {
         return mutation(of: model, modelSchema: modelSchema, where: predicate, type: .delete)
     }
 
+    public static func mutation<M: Model>(of model: M,
+                                          where predicate: QueryPredicate? = nil,
+                                          type: GraphQLMutationType) -> GraphQLRequest<M> {
+        return mutation(of: model, modelSchema: model.schema, where: predicate, type: type)
```
Comment (palpatim) [inline AmplifyPlugins/Core/AWSPluginsCore/Model/GraphQLRequest/GraphQLRequest+Model.swift:155]: nit: remove redundant `return`
```suggestion
        mutation(of: model, modelSchema: model.schema, where: predicate, type: type)
```

**Review by palpatim — APPROVED** [review]
minor style updates, otherwise LGTM

**Review by drochetti — APPROVED** [review]

**Review by drochetti — APPROVED** [review]

**Pull request merged**